docs/19-czym-sa-customcommands.md

```diff
@@ -0,0 +1,21 @@
+# Czym są CustomCommands
```

Comment:
Nie dawno na mastera wszedł mój pull request i trochę się pozmieniało jeśli chodzi o IBotCommand i w tym również Custom Command:
1. Jest to związane ogólnie z komendami, ale można dodać do dokumentacji, że jeśli mamy opcjonalny argument (czyli property z atrybutem Optional) i nie jest to bool ani jakiś typ, który przyjmuje null to powiniśmy uczynić to property - typem, który przyjmuje null. Np: mamy property "public ulong User {get; set;}" jeśli to property ma być Optional to musimy zamienić ulong na ulong? żeby przyjmowało też null. Ponieważ jeśli argument nie został podany przez użytkownika w komendzie to wtedy te property będzie miało wartość null, bo parser automatycznie będzie wstawiał wartość null. Oczywiście to nie jest twarda reguła bo jeśli nie dalibyśmy tego znaku ? to wtedy byłoby 0, i jeśli nam to 0 nie przeszkadza to może ono być (tak jak zrobiono przy warns). Dodajemy znak ? dla typów innych niż bool (bo gdy nie podamy boola to jest false) oraz innych niż typy, które normalnie dopuszczają null (czyli w zasadzie każda klasa np: string dla którego nie trzeba robić string? bo on już przyjmuje null).

2. jeśli chodzi o regexy to ogólnie można już dawać bardziej specyficzne regexy niż tak jak do tej pory, że
zawsze .* . W komendzie messages zamieniłem (?\<User\>.\*) na (?\<User\>\<@!?\d+\>) i dzięki temu działa tak samo
a nawet lepiej (bo mechanizm do rozwiązania "konflikowej sytuacji" rzadziej jest odpalany, ten mechanizm jest odpalany zawsze gdy jest problem czy dana komenda jest default czy custom). Aczkolwiek czasem
może lepiej dać te .* bo jednak, nie zawsze działa to tak jak powinno np: dla update response w OnEvent dałem
.* bo gdybym dał np: [^\s]* to gdyby ktoś podał by "-update response -onevent jeden dwa -message ..." to
komenda by nie została załączona, nawet by nie było żadnego komunikat, a powinno być tak, że powinno do OnEvent
wziąć wartość "jeden", zaś gdyby dał np: .+ to gdybym podał "-update response " to zamiast załączyć komendę i
wyrzucić wyjątek, że podano za mało argumentów to znowu komenda zostanie nie załączona i nie będzie żadnego
komunikatu (dzieje się tak bo żeby komenda została załączona musi być spełniony co najmniej jeden warunek, albo
musi podana komenda zostać załączona z jakimś regexem z bazy lub jej nazwa musi pasować do której z domyślnych
komend, w naszym przypadku ani jedno, ani drugie nie jest spełnione bo regex jest niepoprawny a nazwa podanej
komendy to update a nie update response, bo brane jest tylko pierwsze słowo jako nazwę i z tego wychodzi, że
komenda update nie istnieje i nie jest w ogóle załączane przez nie spełnienie obu tych warunków w dodatku gdy nie jest podana komenda załączona to nie może być komunikatu o błędzie). 
Tych przykładów mógłbym wymyślać wiele, ale gdy decydujemy się zamiast .* dać coś bardziej specjalnego należy wcześniej sprawdzić czy komenda działa dobrze dla każdego przypadku, czyli gdy podamy ją prawidłowo to powinna działać tak jak należy, gdy podamy nieprawidłowy argument to powinien być wyjątek, a gdy nie będzie wymaganego argumentu to też wyrzucić wyjątek. Aczkolwiek zawsze jak nie wiadomo jak zrobić to zawsze można dać po prostu .* to podczas walidacji zostaną wyłapane ewentualne błędy.

3. Podczas pisania kilku regexów do tej samej komendy należy unikać nakładania się tych regexów, nie może być takiej sytuacji, że podana przez użytkownika forma komendy pasuje do kilku regexów jednocześnie, bo wtedy jest brzydki komunikat, że "Podana komenda została dopasowana do więcej niż tylko jednego regexu...". Należy zawsze sprawdzić czy nie ma takiej sytuacji! 
PS. Z resztą mówiłem kiedyś o tym, wcześniej na ogólnym gdy poruszyłem tą kwestie i zastanawiałem się czy nie zrobić specjalnego wyjątka dla tej sytuacji (który faktycznie został stworzony przez ze mnie i jest już na masterze).

4. Ewentualnie można jeszcze o tym napisać: Parser/wykrywacz nie jest idealny, czasem się zdarza, że podano customową komendę w nieprawidłowy sposób, a mimo to nie ma żadnego komunikatu ani odzewu ze strony bota. Jest to normalne dla custom command w szczególności, gdy mamy komendy, których nazwy w customach są wieloczłonowe (np: update response, add role). Np: podamy -update response i wtedy nie ma odzewu, bo nie został spełniony żaden z dwóch warunków, czyli albo musi nazwa komendy pasować do nazwy, któreś z domyślnych metod (w tym przypadku bot pobierze nazwe update, a obecnie nie mamy takiej komendy, więc ten warunek nie jest spełniony) albo musi podana komenda pasować do któregoś regexu w tym przypadku (-update response) nie pasuje do żadnego regexu, bo do tej komendy wymagane są jeszcze argumenty.
Ogólnie jest tak, że najpierw komenda musi być wykryta a potem jest ewentualny komunikat o błędzie. W tym przykładzie podana komenda nie jest wykryta, bo nie spełnia żadnego z wcześniejszych warunków i przez to nie ma wyświetlania błędu.
PS. wiem można to zmienić, żeby tak się nie działo, ale to byłoby trzeba wprowadzać skomplikowane zmiany do kodu i w dodatku i tak nie da się w 100% zapewnić, że gdy customowa komenda została podana, ale w nieprawidłowy sposób, to że zostanie wykryta i będzie komunikat o błędzie.